TRef documentation (#2174)
zio/zio: docs/datatypes/tref.md

---
id: datatypes_tref
title:  "TRef"
---

> A `TRef[A]` is part of the `STM` bundle that represents a transactional reference that can perform reads and writes inside a transactional effect.

Reviewer: ```suggestion
A `TRef[A]` is a mutable reference to an immutable value, which can participate in transactions in STM. The mutable reference can be retrieved and set from within transactions, with strong guarantees for atomicity, consistency, and isolation from other transactions.
```

> `TRef` gives the basic machinery for others `SMT` transactional data types.

Reviewer: ```suggestion
`TRef` provides the low-level machinery to create transactions from modifications of STM memory.
```

`TRef[A]` where `A` is an immutable value.

## Create a TRef

> Creating a `TRef` without committing its value:

Reviewer: ```suggestion
Creating a `TRef` inside a transaction:
```

```scala mdoc:silent
import zio.stm.{STM, TRef}

val tVar: STM[Nothing, TRef[Int]] = TRef.make(10)
```

> Or creating a `TRef` and immediately committing its value to extract out the value:

Reviewer: ```suggestion
Or creating a `TRef` inside a transaction, and immediately committing the transaction, which allows you to store and pass along the reference.
```

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVar: UIO[TRef[Int]] = TRef.makeCommit(10)
```

## Retrieve the value out of a TRef

By retrieving the value in a single transactional statement: 

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = (for {
  tRef <- TRef.make(10)
  value <- tRef.get
} yield value).commit
```

Or on multiple transactional statements:

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = for {
  tRef <- TRef.makeCommit(10)
  value <- tRef.get.commit
} yield value
```

## Set a value to a TRef

By setting a value, it will commit a new value without any relation with the actual value at certain point in time.

Setting a value in a single transactional statement:

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = (for {
  tRef <- TRef.make(10)
  _ <- tRef.set(20)
  nValue <- tRef.get
} yield nValue).commit
```

Or on multiple transactional statement:

> ```scala mdoc:silent

Reviewer: This example is not really realistic because each thing is done in isolation from the others, so there are no atomic guarantees across the set and get.

Author: Sure, you're completely right (also applies for the other examples), for each statement, they're working in isolation. Let me just make sure I'm not wrong, but to have guarantees across the set and get, should I enclosure the statements in a single atomic transaction?
```
for {
  tRef <- TRef(10).commit
  nValue <- STM.atomically {
    for {
      _ <- tRef.set(12)
      nValue <- tRef.get
    } yield nValue
  }
} yield nValue
```
then it gives the guarantees about atomicity of the `set and get`.

import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = for {
  tRef <- TRef.makeCommit(10)
  _ <- tRef.set(20).commit
  value <- tRef.get.commit
} yield value
```

## Update the value of the TRef

> The update function `A => A`, differently from `set`, it cares about the actual value in the `TRef` by performing changes over the value and returning the same type at certain point in time.

Reviewer 2: I'd rephrase to `Update modifies the content of a reference by applying function to it`.

Author: Perfect, let me change to your version.

> The update function `A => A`, differently from `set`, it cares about the actual value in the `TRef` by performing changes over the value and returning the same type at certain point in time.

Reviewer: ```suggestion
The update function `A => A` allows computing a new value for the `TRef` using the old value.
```

Author: Thank you for rephrasing and making even more clarified. (same applies for the other comments too)

Updating the value in a single transactional statement:

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = (for {
  tRef <- TRef.make(10)
  _ <- tRef.update(_ + 20)
  value <- tRef.get
} yield value).commit
```

Or on multiple transactional statements:

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = for {
  tRef <- TRef.makeCommit(10)
  _ <- tRef.update(_ + 20).commit
  value <- tRef.get.commit
} yield value
```

## Modify the value of the TRef

> The modify function `A => (B, A): B` works similar to `update`, but it gives the possibility to perform computation over the actual value and updating to a new value. 

Reviewer: ```suggestion
The modify function `A => (B, A): B` works similar to `update`, but allows extracting some information (the `B`) out of the update operation.
```

---
id: datatypes_tref
title:  "TRef"
---

A `TRef[A]` is part of the `STM` bundle that represents a transactional reference that can perform reads and writes inside a transactional effect.

> `TRef` gives the basic machinery for others `SMT` transactional data types.

Reviewer 2: Fix typo, `SMT` => `STM`.

> `TRef[A]` where `A` is an immutable value.

Reviewer 2: I would remove this sentence.

## Create a TRef

Creating a `TRef` without committing its value:

```scala mdoc:silent
import zio.stm.{STM, TRef}

> val tVar: STM[Nothing, TRef[Int]] = TRef.make(10)

Reviewer 2: `make` has been renamed to `apply`, hence it's `TRef(10)`.

Reviewer: Arg, we should change this back (I am sorry I didn't notice that). In Scala, `TRef.apply` should give you a `TRef`, not a `TRef` inside `STM`. So it's a pervasive ZIO custom that `make` gives you something inside an effect.

Reviewer 2: Gotcha, I'll polish these up asap.

```

Or creating a `TRef` and immediately committing its value to extract out the value:

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVar: UIO[TRef[Int]] = TRef.makeCommit(10)
```

## Retrieve the value out of a TRef

By retrieving the value in a single transactional statement: 

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = (for {

>   tRef <- TRef.make(10)

Reviewer 2: See the comment about renaming.

Author: Sorry, I didn't pull the latest changes. Thanks for noticing. +1

Author: As you made changes to rolling back to `make`, I will leave as it is.

  value <- tRef.get
} yield value).commit
```

Or on multiple transactional statements:

```scala mdoc:silent
import zio.UIO
import zio.stm.TRef

val tVal: UIO[Int] = for {

>   tRef <- TRef.makeCommit(10)

Reviewer 2: I'd consider renaming `makeCommit` method to `committed`. 

@jdegoes wdyt?

Reviewer: It is `make` followed by `commit`. So `makeCommit`. If we kept with `apply` for `make` then `committed` could work well, or even just `commit`, but I am not a fan of that for reasons listed above.

Reviewer 2: Reverted the `apply` renaming in #2179. Please take a look.

  value <- tRef.get.commit
} yield value
```

## Set a value to a TRef

> By setting a value, it will commit a new value without any relation with the actual value at certain point in time.

Reviewer 2: I'd rephrase this to something like `Setting the value overwrites the existing content of a reference`

Author: Nice, I prefer your version, clarifies more.
